### PR TITLE
python310Packages.numpyro: 0.11.0 -> 0.12.1

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Update numpyro to its latest version.
~~We need to wait for https://github.com/NixOS/nixpkgs/pull/246712 to get merged.~~ -> done !

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/funsor/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/funsor/default.nix:77]:
```suggestion
    #   funsor = self.overridePythonAttrs { doCheck = true; };
```

File: pkgs/development/python-modules/funsor/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/funsor/default.nix:54]:
```suggestion
```
those are linters and tests should not break if code quality is not correct

File: pkgs/development/python-modules/funsor/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/funsor/default.nix:34]:
```suggestion
      repo = "funsor";
```

File: pkgs/development/python-modules/funsor/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/funsor/default.nix]: Ok !

**Comment by GaetanLepage** [comment]
Thank you @SuperSandro2000 for the review.
Currently, some tests are still failing for `funsor`.
I will keep investigating and leave this PR as draft in the meantime.

File: pkgs/development/python-modules/funsor/default.nix
Comment (samuela) [inline pkgs/development/python-modules/funsor/default.nix:79]: why is this let-in necessary?

File: pkgs/development/python-modules/funsor/default.nix
Comment (samuela) [inline pkgs/development/python-modules/funsor/default.nix:31]:
```suggestion
      repo = pname;
```

File: pkgs/development/python-modules/funsor/default.nix
Comment (samuela) [inline pkgs/development/python-modules/funsor/default.nix:62]: in that case let's report upstream and link to the issue in a code comment here

File: pkgs/development/python-modules/funsor/default.nix
Comment (samuela) [inline pkgs/development/python-modules/funsor/default.nix:59]: why is this necessary? isn't this the default?

**Review by samuela — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/funsor/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/funsor/default.nix:31]: This practice has been discouraged by maintainers recently to avoid errors if the package gets renamed.

File: pkgs/development/python-modules/funsor/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/funsor/default.nix]: You are right ! Removed it :)

**Comment by K900** [comment]
FWIW I'm not really familiar with scientific Python stuff in Nixpkgs so maybe don't tag me on it :)

**Review by samuela — APPROVED** [review]
diff LGTM happy to merge once nixpkgs-review comes back clean

EDIT: although it would still be nice to resolve https://github.com/NixOS/nixpkgs/pull/246761#discussion_r1297566449 first

File: pkgs/development/python-modules/funsor/default.nix
Comment (samuela) [reply pkgs/development/python-modules/funsor/default.nix]: fair enough

File: pkgs/development/python-modules/funsor/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/funsor/default.nix]: Done :)

**Comment by GaetanLepage** [comment]
I managed to successfully test `funsor` using the included `Makefile`.
I only disabled the linting checks and the tests that require internet access.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 246761` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.bambi</li>
    <li>python310Packages.bambi.dist</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.pymc.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>python310Packages.arviz</li>
    <li>python310Packages.arviz.dist</li>
    <li>python310Packages.funsor</li>
    <li>python310Packages.funsor.dist</li>
    <li>python310Packages.numpyro</li>
    <li>python310Packages.numpyro.dist</li>
    <li>python311Packages.funsor</li>
    <li>python311Packages.funsor.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Failures of `bambi` and `pymc` are unrelated to this update.

File: pkgs/development/python-modules/funsor/patch-makefile-for-tests.patch
Comment (samuela) [inline pkgs/development/python-modules/funsor/patch-makefile-for-tests.patch:1]: I prefer using `substituteInPlace` in `postPatch` instead of patch files where possible since patch files tend to be more frustrating to modify when updating the package in the future, eg. upstream modifies the Makefile. But it's not a critical choice one way or the other

**Comment by samuela** [comment]
LGTM I'll plan on merging once ofborg finishes

**Review by dotlambda — COMMENTED** [review]
`pythonImportsCheck` for funsor would be nice

**Comment by GaetanLepage** [comment]
> `pythonImportsCheck` for funsor would be nice

Nice catch ! Fixed :)

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 246761` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.bambi</li>
    <li>python310Packages.bambi.dist</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.pymc.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>python310Packages.arviz</li>
    <li>python310Packages.arviz.dist</li>
    <li>python310Packages.funsor</li>
    <li>python310Packages.funsor.dist</li>
    <li>python310Packages.numpyro</li>
    <li>python310Packages.numpyro.dist</li>
    <li>python311Packages.funsor</li>
    <li>python311Packages.funsor.dist</li>
  </ul>
</details>

**Pull request merged**
